module/zfs/txg.c

```diff
@@ -566,10 +555,24 @@ txg_sync_thread(void *arg)
 		 * prompting it to do so if necessary.
 		 */
 		while (!tx->tx_exiting && !txg_has_quiesced_to_sync(dp)) {
-			if (tx->tx_quiesce_txg_waiting < tx->tx_open_txg+1)
-				tx->tx_quiesce_txg_waiting = tx->tx_open_txg+1;
-			cv_broadcast(&tx->tx_quiesce_more_cv);
-			txg_thread_wait(tx, &cpr, &tx->tx_quiesce_done_cv, 0);
+			if (txg_is_quiescing(dp)) {
+				txg_thread_wait(tx, &cpr,
+				    &tx->tx_quiesce_done_cv, 0);
+				continue;
+			}
+			txg = tx->tx_open_txg;
+			tx->tx_quiescing_txg = txg;
+
+			mutex_exit(&tx->tx_sync_lock);
+			txg_quiesce(dp, txg);
+			mutex_enter(&tx->tx_sync_lock);
+
+			dprintf("quiesce done, txg %llu\n", (u_longlong_t)txg);
+			tx->tx_quiescing_txg = 0;
+			tx->tx_quiesced_txg = txg;
+			DTRACE_PROBE2(txg__quiesced, dsl_pool_t *, dp,
+			    uint64_t, txg);
+			cv_broadcast(&tx->tx_quiesce_done_cv);
```

Comment:
The existing code was surprisingly tricky to get right.  The proposed change here makes the code more complicated (we could either do the quiesce from the quiesce thread or from the sync thread).  Can you share some performance results to motivate the added complexity?

Reply:
The quiescing process is so fast that it even costs less than context switch between threads. My intention is it may not be worth to quiesce txgs on the separated thread. 
Is the performance difference noticeable. Sadly No. Quiescing/Syncing only happens once every a couple of seconds. Saving miro-seconds out of seconds is not measurable.

This portion of code may be left as it is on current "openzfs/zfs". 
Or do you suggest to remove the txg_quiesce_thread all together? Maybe for the future Pull Request.

Reply:
I would suggest removing these changes from this PR, and perhaps look at removing the txg_quiesce_thread alltogether in a future PR.  Since the performance implications are not important, I would make those changes with an eye to improving code clarity / design.

Reply:
Just re-pushed the commit.
```shell
                while (!tx->tx_exiting && !txg_has_quiesced_to_sync(dp)) {
+                       if (txg_is_quiescing(dp)) {
+                               txg_thread_wait(tx, &cpr,
+                                   &tx->tx_quiesce_done_cv, 0);
+                               continue;
+                       }
                        if (tx->tx_quiesce_txg_waiting < tx->tx_open_txg+1)
                                tx->tx_quiesce_txg_waiting = tx->tx_open_txg+1;
                        cv_broadcast(&tx->tx_quiesce_more_cv);
                        txg_thread_wait(tx, &cpr, &tx->tx_quiesce_done_cv, 0);
                }
```

The reason I want to keep that portion in is that I do find some very rare race condition that tx_open_txg changed between "if (tx->tx_quiesce_txg_waiting < tx->tx_open_txg+1)" and "tx->tx_quiesce_txg_waiting = tx->tx_open_txg+1;".